text/0000-ssr-stylesheet-deduplication.md

```diff
@@ -0,0 +1,177 @@
+### API shape
+
+A new property bag is added to the `renderComponent` API of `@lwc/ssr-runtime`.
+
+```ts
+function renderComponent(
+    tagName: string,
+    Constructor: LightningElement,
+    props: any,
+    // The below is new
+    {
+        renderStylesheet(contents: string) {
+            return {
+                href: '/some-file.css', // required
+                crossorigin: 'anonymous', // optional
+                fetchpriority: 'high', // optional
+                integrity: 'sha384-<hash>', // optional
+                referrerpolicy: 'no-referrer', // optional
+            }
+        }
+    }
+)
+```
+
+The API allows for any number of these attributes to be specified in the returned object. The only required property is `href`. (`rel=stylesheet` is implied.)
+
+The `renderStylesheet` function may also return a falsy value, in which case it is assumed that the `<style>` should _not_ be transformed into a `<link rel=stylesheet>`. This allows for flexibility in skipping the transformation if, for example, the stylesheet contents are sufficiently small.
+
+The `renderStylesheet` function is passed in as part of a second options bag after the `props` passed in to `renderComponent`. All props are optional.
+
+### Prior art
+
+#### Web component frameworks
+
+Lit has [toyed with the idea](https://github.com/lit/lit/issues/4357) of deduplicating styles, but not committed to anything. The two high-level ideas they have explored are the `<link rel=stylesheet>` solution and the "inline web component" solution.
+
+FAST has implemented a ["styled component"](https://github.com/microsoft/fast/pull/5600) approach relying on an inline web component to copy styles from a global source into individual components. (Notably this approach has led to [Flash of Unstyled Content issues](https://github.com/microsoft/fast/issues/5906).)
+
+The Microsoft Edge team has [an explainer](https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/ShadowDOM/explainer.md) proposing a solution to this problem at the level of HTML, but this is still far from being available cross-browser. And when it is, it is unclear how backwards compatibility with older browsers should work.
+
+#### Non-web-component frameworks
+
+Most non-web-component using frameworks (such as Vue, Svelte, and React) have a system where styles can be extracted from compiled components and are assumed to be hoisted to the document `<head>`. In fact, in Vite, that is what this does:
+
+```js
+import './stylesheet.css' // auto-hoists
+```
+
+Needless to say, this is insufficient for our use case since LWC has first-class support for shadow DOM components.
+
+## Drawbacks
+
+This solution will add significant complexity both to the LWC engine (which will have to deal with supporting the new API in SSR, as well as dealing with hydration and mismatch detection on the client) and on the consumer side (which will have to do the bulk of the work of managing and hosting the stylesheets themselves).

```

Comment:
Can you explain why this change increases the possibility of hydration mismatch?

Reply:
Right now the hydration mechanism for `<style>`s is the same as any other DOM element. Whereas if we render `<link>`s then we will need to instruct hydration with something like "If you see a `<style>` with such-and-such text content (e.g. a checksum), then consider it the same as a `<link>` with such-and-such attributes."